### DISCUSSION TITLE
Configuration CC V3+: Finding the first configuration parameter with getProperties(0)

### DISCUSSION THREAD

**Discussion by AlCalzone** [zwave-js/zwave-js#1414] [General] [answered]
The specs mention that querying the non-existant configuration parameter #0 should result in a reply with the first defined configuration parameter of a device. The reality is different:

**Heatit Z-TRM2:** no next parameter
```
19:54:34.490 SERIAL » 0x010b00130904700e000025d968                                        (13 bytes)
19:54:34.491 DRIVER » [Node 009] [REQ] [SendData]
                      │ transmit options: 0x25
                      │ callback id:      217
                      └─[ConfigurationCCPropertiesGet]
                          parameter #: 0
...
19:54:34.787 SERIAL « 0x0110000400090a700f000001000000000197                              (18 bytes)
19:54:34.789 SERIAL » [ACK]                                                                   (0x06)
19:54:34.790 DRIVER « [Node 009] [REQ] [ApplicationCommand]
                      └─[ConfigurationCCPropertiesReport]
                          parameter #:   0
                          next param #:  0
                          value size:    1
                          value format:  SignedInteger
                          min value:     0
                          max value:     0
                          default value: 0
```

**Heatit Z-TRM2fx:** supposedly the improved version, but that replies with an invalid command (payload is just 3 zero-bytes)
```
14:43:30.584 CNTRLR » [Node 047] finding first configuration parameter...
14:43:30.586 SERIAL » 0x010b00132f04700e000025dd4a                                        (13 bytes)
14:43:30.586 DRIVER » [Node 047] [REQ] [SendData]
                      │ transmit options: 0x25
                      │ callback id:      221
                      └─[ConfigurationCCPropertiesGet]
                          parameter #: 0
...
14:43:30.625 SERIAL « 0x010b0004002f05700f000000a5                                        (13 bytes)
```

@hanskroner:
I'm working around this by trying parameter #1 if we get this invalid 3-zeroes response or a ConfigurationCCPropertiesReport with next param # 0.
I'm wondering if there's a better way to do this?

**Comment by Xrlls** [comment, marked as answer]
As you wrote in your initial post, the requirement, CC:0070.01.05.12.002, is a "should", not a must, and as you have found, not all devices fulfill the requirement. Specifically for old devices, there may not be any discovery mechanism at all, except for trying each config parameter, and see if it results in a response: Many of the older devices relied on the configuration parameters being specified on paper in the user guide, with no over the air discovery mechanisms, telling which parameters exists, what the valid values are etc.

**Reply by AlCalzone** [reply]
> Specifically for old devices, there may not be any discovery mechanism at all

We're relying on device configuration files for that, as scanning 255 possible parameters for each device would cause too much traffic.

> is a "should", not a must, and as you have found, not all devices fulfill the requirement

So if we're looking at devices with Configuration CC V3+, there is really no better way, correct? If the device doesn't tell us the first parameter after querying #0 and #1, we're better off using the configuration file information instead?

**Reply by Xrlls** [reply]
Agreed!
